Undefined locale parameters in service definitions

**Sylius version affected**: -
**Description**  
Some services are configured using the `%locale%` parameter, which is undefined by default, so when trying to apply an auto-generated Flex recipe, an error occurs:

```
- Configuring sylius/resource-bundle (>=dev-***): From auto-generated recipe
Executing script cache:clear [KO]
 [KO]
Script cache:clear returned with error code 1
!!  
!!  In DefinitionErrorExceptionPass.php line 51:
!!                                                                                 
!!    You have requested a non-existent parameter "locale". Did you mean one of t  
!!    hese: "kernel.default_locale", "kernel.enabled_locales"?                     
!!                                                                                 
!!  
!!  
Script @auto-scripts was called via post-update-cmd
Error: Process completed with exit code 1.
```

However, if a ready-made recipe is used, the installation is successful:
```
- Configuring sylius/resource-bundle (>=1.12): From github.com/symfony/recipes-contrib:main
Executing script cache:clear [OK]
Executing script assets:install public [OK]
```

This is because the recipe adds the `locale` parameter to the dependency container:
https://github.com/symfony/recipes-contrib/blob/afcbbbca3fac2dde0b7f13a498b9cebe2699dab4/sylius/resource-bundle/1.12/manifest.json#L3

As a result, CI errors occur: https://github.com/Sylius/SyliusResourceBundle/pull/1004#issuecomment-2940566162

**Steps to reproduce**  
1. Create project using `symfony/skeleton`
2. Install `sylius/resource-bundle` (using the auto-generated recipe)
3. Review the logs for errors

**Possible Solution**  
To ensure that tests using `symfony/skeleton` (PR #1014) pass even when using the auto-generated Flex recipe and for better compatibility, you should use the kernel locale configuration parameters:

https://symfony.com/doc/current/reference/configuration/kernel.html#kernel-enabled-locales
https://symfony.com/doc/current/reference/configuration/kernel.html#kernel-default-locale

**Additional context**  
Related issue: https://github.com/Sylius/SyliusResourceBundle/issues/244

--- a/src/Bundle/SyliusResourceBundle.php
+++ b/src/Bundle/SyliusResourceBundle.php
@@ -30,6 +30,7 @@ use Sylius\Bundle\ResourceBundle\DependencyInjection\Compiler\UnregisterHateoasD
 use Sylius\Bundle\ResourceBundle\DependencyInjection\Compiler\WinzouStateMachinePass;
 use Sylius\Bundle\ResourceBundle\DependencyInjection\PagerfantaExtension;
 use Sylius\Resource\Symfony\DependencyInjection\Compiler\DisableMetadataCachePass;
+use Sylius\Resource\Symfony\DependencyInjection\Compiler\FallbackToKernelDefaultLocalePass;
 use Symfony\Component\DependencyInjection\Compiler\PassConfig;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\HttpKernel\Bundle\Bundle;
@@ -50,6 +51,7 @@ final class SyliusResourceBundle extends Bundle
 
         $container->addCompilerPass(new CsrfTokenManagerPass());
         $container->addCompilerPass(new DisableMetadataCachePass());
+        $container->addCompilerPass(new FallbackToKernelDefaultLocalePass());
         $container->addCompilerPass(new DoctrineContainerRepositoryFactoryPass());
         $container->addCompilerPass(new DoctrineTargetEntitiesResolverPass(new TargetEntitiesResolver()), PassConfig::TYPE_BEFORE_OPTIMIZATION, 1);
         $container->addCompilerPass(new RegisterFormBuilderPass());

--- a/src/Component/src/Symfony/DependencyInjection/Compiler/FallbackToKernelDefaultLocalePass.php
+++ b/src/Component/src/Symfony/DependencyInjection/Compiler/FallbackToKernelDefaultLocalePass.php
@@ -1,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Resource\Symfony\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * @internal
+ */
+final class FallbackToKernelDefaultLocalePass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (
+            $container->hasParameter('locale') ||
+            !$container->hasParameter('kernel.default_locale')
+        ) {
+            return;
+        }
+
+        /** @var string $locale */
+        $locale = $container->getParameter('kernel.default_locale');
+
+        $this->replaceLocaleInFlashHelper($container, $locale);
+        $this->replaceLocaleProvider($container, $locale);
+    }
+
+    private function replaceLocaleInFlashHelper(ContainerBuilder $container, string $locale): void
+    {
+        if (!$container->hasDefinition('sylius.resource_controller.flash_helper')) {
+            return;
+        }
+
+        $flashHelper = $container->getDefinition('sylius.resource_controller.flash_helper');
+        $flashHelper->replaceArgument(2, $locale);
+    }
+
+    private function replaceLocaleProvider(ContainerBuilder $container, string $locale): void
+    {
+        if (!$container->hasDefinition('sylius.translation_locale_provider.immutable')) {
+            return;
+        }
+
+        $localeProvider = $container->getDefinition('sylius.translation_locale_provider.immutable');
+        $localeProvider->replaceArgument(0, [$locale]);
+        $localeProvider->replaceArgument(1, $locale);
+    }
+}

--- a/src/Component/tests/Symfony/DependencyInjection/Compiler/DisableMetadataCachePassTest.php
+++ b/src/Component/tests/Symfony/DependencyInjection/Compiler/DisableMetadataCachePassTest.php
@@ -11,7 +11,7 @@
 
 declare(strict_types=1);
 
-namespace Sylius\Component\Resource\tests\Symfony\DependencyInjection\Compiler;
+namespace Sylius\Resource\Tests\Symfony\DependencyInjection\Compiler;
 
 use Matthias\SymfonyDependencyInjectionTest\PhpUnit\AbstractCompilerPassTestCase;
 use Sylius\Resource\Symfony\DependencyInjection\Compiler\DisableMetadataCachePass;

--- a/src/Component/tests/Symfony/DependencyInjection/Compiler/FallbackToKernelDefaultLocalePassTest.php
+++ b/src/Component/tests/Symfony/DependencyInjection/Compiler/FallbackToKernelDefaultLocalePassTest.php
@@ -1,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Resource\Tests\Symfony\DependencyInjection\Compiler;
+
+use Matthias\SymfonyDependencyInjectionTest\PhpUnit\AbstractCompilerPassTestCase;
+use PHPUnit\Framework\Attributes\Test;
+use Sylius\Bundle\ResourceBundle\Controller\FlashHelper;
+use Sylius\Resource\Symfony\DependencyInjection\Compiler\FallbackToKernelDefaultLocalePass;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+
+final class FallbackToKernelDefaultLocalePassTest extends AbstractCompilerPassTestCase
+{
+    /** @test */
+    public function it_does_nothing_if_locale_parameter_has_been_defined(): void
+    {
+        $this->setParameter('locale', 'en_US');
+        $this->setDefinition('sylius.resource_controller.flash_helper', new Definition());
+
+        $this->compile();
+
+        $this->assertContainerBuilderHasService('sylius.resource_controller.flash_helper');
+    }
+
+    /** @test */
+    public function it_replaces_locale_argument_in_flash_helper(): void
+    {
+        $flashHelperDefinition = (new Definition(FlashHelper::class))->setArguments([
+            null,
+            null,
+            null,
+        ]);
+
+        $this->setParameter('kernel.default_locale', 'en_US');
+        $this->setDefinition('sylius.resource_controller.flash_helper', $flashHelperDefinition);
+
+        $this->compile();
+
+        $this->assertContainerBuilderHasService('sylius.resource_controller.flash_helper');
+        $this->assertContainerBuilderHasServiceDefinitionWithArgument('sylius.resource_controller.flash_helper', 2, 'en_US');
+    }
+
+    /** @test */
+    public function it_replaces_locale_argument_in_translation_locale_provider(): void
+    {
+        $translationLocaleProviderDefinition = (new Definition(FlashHelper::class))->setArguments([
+            [],
+            null,
+        ]);
+
+        $this->setParameter('kernel.default_locale', 'en_US');
+        $this->setDefinition('sylius.translation_locale_provider.immutable', $translationLocaleProviderDefinition);
+
+        $this->compile();
+
+        $this->assertContainerBuilderHasService('sylius.translation_locale_provider.immutable');
+        $this->assertContainerBuilderHasServiceDefinitionWithArgument('sylius.translation_locale_provider.immutable', 0, ['en_US']);
+        $this->assertContainerBuilderHasServiceDefinitionWithArgument('sylius.translation_locale_provider.immutable', 1, 'en_US');
+    }
+
+    protected function registerCompilerPass(ContainerBuilder $container): void
+    {
+        $container->addCompilerPass(new FallbackToKernelDefaultLocalePass());
+    }
+}

--- a/tests/Application/config/packages/framework.yaml
+++ b/tests/Application/config/packages/framework.yaml
@@ -2,11 +2,11 @@ framework:
     assets: false
     translator:
         default_path: '%kernel.project_dir%/translations'
-        fallbacks: ["%locale%"]
+        fallbacks: ["en_US"]
     secret: "%secret%"
     form: ~
     csrf_protection: true
-    default_locale: "%locale%"
+    default_locale: "en_US"
     session:
         handler_id: ~
         storage_factory_id: session.storage.factory.mock_file

--- a/tests/Application/config/services.yaml
+++ b/tests/Application/config/services.yaml
@@ -5,7 +5,6 @@ parameters:
     database_driver: pdo_sqlite
     database_path: "%kernel.project_dir%/config/db.sql"
 
-    locale: en_US
     secret: "Three can keep a secret, if two of them are dead."
 
 services:

